[FP8] Enable FP16 output support for torch scaled_mm when using CUTLASS on SM90

oyye opened the pull request:
Summary: NVIDIA uses CUTLASS for row-wise scaling prior to cuBLAS version 12.9. This change enables support for FP16 data type for both bias and output when using CUTLASS.

Test Plan:
pytest -svv test/test_scaled_matmul_cuda.py

Test results on cuda-12.4:
```
test/test_scaled_matmul_cuda.py::TestFP8MatmulCUDA::test_scaled_mm_vs_emulated_row_wise_bfloat16_cuda PASSED [0.0022s]
test/test_scaled_matmul_cuda.py::TestFP8MatmulCUDA::test_scaled_mm_vs_emulated_row_wise_float16_cuda PASSED [0.0023s]
test/test_scaled_matmul_cuda.py::TestFP8MatmulCUDA::test_scaled_mm_vs_emulated_row_wise_float32_cuda SKIPPED [0.0005s]
======================= 51 passed, 516 skipped in 5.26s ========================
```

Test results on cuda-12.9:
```
test/test_scaled_matmul_cuda.py::TestFP8MatmulCUDA::test_scaled_mm_vs_emulated_row_wise_bfloat16_cuda PASSED [0.0046s]
test/test_scaled_matmul_cuda.py::TestFP8MatmulCUDA::test_scaled_mm_vs_emulated_row_wise_float16_cuda PASSED [0.0040s]
test/test_scaled_matmul_cuda.py::TestFP8MatmulCUDA::test_scaled_mm_vs_emulated_row_wise_float32_cuda PASSED [0.0038s]
======================= 70 passed, 482 skipped in 5.88s ========================
```

Reviewed By: pranavsharma, RandySheriff

Differential Revision: D84169910

slayton58 reviewed (commented):
Changes look good, functionality should be made equivalent via. `_scaled_mm_v2` (which does already use `_scaled_rowwise_rowwise` so changes should be minor.

oyye commented on test/test_scaled_matmul_cuda.py:
```diff
@@ -1188,12 +1188,23 @@ def test_scaled_mm_vs_emulated_row_wise(self, base_dtype):
+                raise unittest.SkipTest("hipblaslt rowwise _scaled_mm only supports Float16")
+            if torch.cuda.get_device_capability() < (9, 0):
+                raise unittest.SkipTest("Need sm90+ for row-wise fp8 w/ cuBLAS")
+            use_wrap = False
```
Sure, I will update the v2 api.

oyye committed:
[FP8] Enable FP16 output support for torch scaled_mm when using CUTLASS on SM90 (#166744)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/166744

NVIDIA uses CUTLASS for row-wise scaling prior to cuBLAS version 12.9. This change enables support for FP16 data type for both bias and output when using CUTLASS.

Test Plan:
pytest -svv test/test_scaled_matmul_cuda.py

Test results on cuda-12.4:
```
test/test_scaled_matmul_cuda.py::TestFP8MatmulCUDA::test_scaled_mm_vs_emulated_row_wise_bfloat16_cuda PASSED [0.0022s]
test/test_scaled_matmul_cuda.py::TestFP8MatmulCUDA::test_scaled_mm_vs_emulated_row_wise_float16_cuda PASSED [0.0024s]
test/test_scaled_matmul_cuda.py::TestFP8MatmulCUDA::test_scaled_mm_vs_emulated_row_wise_float32_cuda SKIPPED [0.0005s]
======================= 51 passed, 516 skipped in 4.94s ========================
```

Test results on cuda-12.9:
```
test/test_scaled_matmul_cuda.py::TestFP8MatmulCUDA::test_scaled_mm_vs_emulated_row_wise_bfloat16_cuda PASSED [0.0048s]
test/test_scaled_matmul_cuda.py::TestFP8MatmulCUDA::test_scaled_mm_vs_emulated_row_wise_float16_cuda PASSED [0.0041s]
test/test_scaled_matmul_cuda.py::TestFP8MatmulCUDA::test_scaled_mm_vs_emulated_row_wise_float32_cuda PASSED [0.0038s]
======================= 70 passed, 482 skipped in 4.57s ========================
```

Reviewed By: pranavsharma, RandySheriff

Differential Revision: D84169910

oyye commented on test/test_scaled_matmul_cuda.py:
```diff
@@ -1208,13 +1219,17 @@ def test():
             out_emulated = mm_float8_emulated(
                 x_fp8, x_scales, y_fp8, y_scales, output_dtype
             )
+            if bias is not None:
```
Updated with adding bias parameter for mm_float8_emulated.

slayton58 reviewed (approved):
LGTM
